src/result.cpp

```diff
@@ -175,13 +181,38 @@ read_embedded_file_entry(CacheEntryReader& reader,
       }
       uint8_t buf[READ_BUFFER_SIZE];
       size_t remain = file_len;
-      while (remain > 0) {
+
+      // data which has to be read in its entirety because it has to be
+      // modified before it can be written out.
+      if (FileType(type) == FileType::compile_command) {
         size_t n = std::min(remain, sizeof(buf));
         reader.read(buf, n);
-        if (fwrite(buf, n, 1, subfile.get()) != 1) {
+
+        remain -= n;
+
+        // 64k ought to be enough for anybody's ccmd text file
+        if (remain > 0) {
+          throw Error("stored .ccmd file is too big");
+        }
```

Comment:
Relying on READ_BUFFER_SIZE remaining larger than any feasible json fragment. Maybe add a static assert. Otherwise would need to duplicate the while loop from below.